<a href="https://colab.research.google.com/github/igoekce/netflixrecommender/blob/Data-Preprocessing/Data_Preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import

In [ ]:
# import librairies
import networkx as nx
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import math as math
import time 
plt.style.use('seaborn')
plt.rcParams['figure.figsize'] = [14,14]

filepath = 'https://raw.githubusercontent.com/igoekce/netflixrecommender/master/netflix_titles.csv'

df = pd.read_csv(filepath)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6234 entries, 0 to 6233
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       6234 non-null   int64 
 1   type          6234 non-null   object
 2   title         6234 non-null   object
 3   director      4265 non-null   object
 4   cast          5664 non-null   object
 5   country       5758 non-null   object
 6   date_added    6223 non-null   object
 7   release_year  6234 non-null   int64 
 8   rating        6224 non-null   object
 9   duration      6234 non-null   object
 10  listed_in     6234 non-null   object
 11  description   6234 non-null   object
dtypes: int64(2), object(10)
memory usage: 584.6+ KB


In [ ]:
#Erzeugung eines Datensatzes mit 20% der Daten
df = df.sample(int(df.shape[0]*0.2))
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1246 entries, 5599 to 1259
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       1246 non-null   int64 
 1   type          1246 non-null   object
 2   title         1246 non-null   object
 3   director      845 non-null    object
 4   cast          1127 non-null   object
 5   country       1144 non-null   object
 6   date_added    1242 non-null   object
 7   release_year  1246 non-null   int64 
 8   rating        1243 non-null   object
 9   duration      1246 non-null   object
 10  listed_in     1246 non-null   object
 11  description   1246 non-null   object
dtypes: int64(2), object(10)
memory usage: 126.5+ KB


In [ ]:
df

show_id  ...                                        description
5599  80216752  ...  This enlightening series from Vox digs into a ...
1692  70259801  ...  A disgraced Secret Service agent must come to ...
5955  80091742  ...  In a docuseries set at one of NCAA football's ...
3381  81144867  ...  A reserved high school freshman acts out when ...
1710  80996626  ...  True and her friends are making music – and th...
...        ...  ...                                                ...
3856  81026327  ...  While she successfully advocates for her femal...
4438  80081754  ...  In the time of dinosaurs, the ancient and powe...
5063  60010514  ...  When an aging flight attendant's caught smuggl...
1029  80009431  ...  The Oscar-nominated true story of the rangers ...
1259  80200267  ...  Ranging from absurd to profound, these Western...

[1246 rows x 12 columns]

1246 tilles = missing **director**, **cast**, **country**, date_added, rating

#separating columns multi text

In [ ]:
df_sep = df.copy()
df_sep = df_sep.dropna()

In [ ]:
df_sep

show_id  ...                                        description
1692  70259801  ...  A disgraced Secret Service agent must come to ...
3381  81144867  ...  A reserved high school freshman acts out when ...
5417  80015280  ...  Teen witch Bibi and horse-loving Tina face tro...
1753  70153178  ...  Mumbai's top mob boss rules the underworld wit...
972   70153553  ...  A young Kashmiri man sets out to escape to Pak...
...        ...  ...                                                ...
571   70082633  ...  Siddharth dismisses warnings that his palace m...
2284  80134438  ...  Bored with her life and inspired by a friend's...
3856  81026327  ...  While she successfully advocates for her femal...
5063  60010514  ...  When an aging flight attendant's caught smuggl...
1259  80200267  ...  Ranging from absurd to profound, these Western...

[739 rows x 12 columns]

In [ ]:
def splitDataFrameList(df,target_column,separator):
    ''' df = dataframe to split,
    target_column = the column containing the values to split
    separator = the symbol used to perform the split
    returns: a dataframe with each entry for the target column separated, with each element moved into a new row. 
    The values in the other columns are duplicated across the newly divided rows.
    '''
    def splitListToRows(row,row_accumulator,target_column,separator):
        split_row = row[target_column].split(separator)
        for s in split_row:
            new_row = row.to_dict()
            new_row[target_column] = s
            row_accumulator.append(new_row)
    new_rows = []
    df.apply(splitListToRows,axis=1,args = (new_rows,target_column,separator))
    new_df = pd.DataFrame(new_rows)
    return new_df

In [ ]:
df_sep = splitDataFrameList(df_sep,'cast',',')
df_sep = splitDataFrameList(df_sep,'country',',')
df_sep = splitDataFrameList(df_sep,'listed_in',',')
df_sep

show_id  ...                                        description
0      70259801  ...  A disgraced Secret Service agent must come to ...
1      70259801  ...  A disgraced Secret Service agent must come to ...
2      70259801  ...  A disgraced Secret Service agent must come to ...
3      70259801  ...  A disgraced Secret Service agent must come to ...
4      70259801  ...  A disgraced Secret Service agent must come to ...
...         ...  ...                                                ...
17283  80200267  ...  Ranging from absurd to profound, these Western...
17284  80200267  ...  Ranging from absurd to profound, these Western...
17285  80200267  ...  Ranging from absurd to profound, these Western...
17286  80200267  ...  Ranging from absurd to profound, these Western...
17287  80200267  ...  Ranging from absurd to profound, these Western...

[17288 rows x 12 columns]

#Function generating G for model
source = title

target = could be 2 to 7 features

In [ ]:
def G_generator(features):
  df1 = pd.DataFrame(columns = ['title', 'value']) 
  for feature in features:
    print(feature)
    df_temp = df_sep[['title', feature]].copy()
    df_temp.columns = ['title', 'value']
    df1 = pd.concat([df1, df_temp], axis=0)
  return df1
#  G = nx.from_pandas_edgelist(df1, source='title', target='value')

#abfangen von leeren Zellen?

In [ ]:
print('possible cols w/o title:', df_sep.columns)

possible cols w/o title: Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description'],
      dtype='object')


In [ ]:
#PLEASE INSERT YOUR COLS
features = ['rating', 'cast', 'release_year', 'duration']
features = ['rating', 'cast']

In [ ]:
df_model = G_generator(features)

rating
cast


title             value
0                Olympus Has Fallen                 R
1                Olympus Has Fallen                 R
2                Olympus Has Fallen                 R
3                Olympus Has Fallen                 R
4                Olympus Has Fallen                 R
...                             ...               ...
17283  The Ballad of Buster Scruggs         Tom Waits
17284  The Ballad of Buster Scruggs         Zoe Kazan
17285  The Ballad of Buster Scruggs         Zoe Kazan
17286  The Ballad of Buster Scruggs   Brendan Gleeson
17287  The Ballad of Buster Scruggs   Brendan Gleeson

[34576 rows x 2 columns]

In [ ]:
G_generator(['director', 'cast'])

director
cast


title             value
0                Olympus Has Fallen     Antoine Fuqua
1                Olympus Has Fallen     Antoine Fuqua
2                Olympus Has Fallen     Antoine Fuqua
3                Olympus Has Fallen     Antoine Fuqua
4                Olympus Has Fallen     Antoine Fuqua
...                             ...               ...
17283  The Ballad of Buster Scruggs         Tom Waits
17284  The Ballad of Buster Scruggs         Zoe Kazan
17285  The Ballad of Buster Scruggs         Zoe Kazan
17286  The Ballad of Buster Scruggs   Brendan Gleeson
17287  The Ballad of Buster Scruggs   Brendan Gleeson

[34576 rows x 2 columns]

#DF Preparation (not needed anymore)
Available Nodes:
'directors', 'actors' , 'countries', 'release_year', 'rating', 'duration', 'categories'

In [ ]:
# convert columns "director, listed_in, cast and country" in columns that contain a real list
# the strip function is applied on the elements
# if the value is NaN, the new column contains a empty list []
df['directors'] = df['director'].apply(lambda l: [] if pd.isna(l) else [i.strip() for i in l.split(",")])
df['categories'] = df['listed_in'].apply(lambda l: [] if pd.isna(l) else [i.strip() for i in l.split(",")])
df['actors'] = df['cast'].apply(lambda l: [] if pd.isna(l) else [i.strip() for i in l.split(",")])
df['countries'] = df['country'].apply(lambda l: [] if pd.isna(l) else [i.strip() for i in l.split(",")])

df.head()

show_id  ...                countries
4889  80134952  ...  [United States, Canada]
1820  81069255  ...         [United Kingdom]
1351  80216003  ...          [United States]
519   80235268  ...        [Spain, Portugal]
5219  80185032  ...          [United States]

[5 rows x 16 columns]

In [ ]:
df1 = pd.DataFrame( np.concatenate( (
    df[['title', 'directors']].values, 
    df[['title', 'actors']].values, 
    df[['title', 'countries']].values, 
    df[['title', 'release_year']].values, 
    df[['title', 'rating']].values, 
    df[['title', 'duration']].values, 
    df[['title', 'categories']].values, 
    ), axis=0 ) )
df1.columns = ['title', 'value']
df1

title                                      value
0                                  Bad Rap                            [Salima Koroma]
1           John & Yoko: Above Us Only Sky                          [Michael Epstein]
2                           Apache Warrior        [David Salzberg, Christian Tureaud]
3                                     Jefe                          [Sergio Barrejón]
4         Behind the Curtain: Todrick Hall                 [Katherine Fairfax Wright]
...                                    ...                                        ...
8717                 Chris Rock: Tamborine                          [Stand-Up Comedy]
8718             Harry Benson: Shoot First                            [Documentaries]
8719                                Lolita  [Classic Movies, Dramas, Romantic Movies]
8720  Thelonious Monk: Straight, No Chaser          [Documentaries, Music & Musicals]
8721                             New Money             [Dramas, International Movies]

[8722 rows x 2 columns]

# Ideen

Ideen:

- Es gibt deutlich mehr Filme als Serien, d.h. ggf. prüfen, ob der eingegebene Name ein Film oder eine Serie ist (oder ob er überhaupt auftaucht).
- Filmtitel ist **nicht** unique. Das sollten wir klären. Die Duplikate führen beim Grafen zu Connections, welche eigentlich nicht da sein sollten.
- ca. 1/3 der Directors fehlen (kriegen wir die woanders her? z.B. IMDB)
- Cast müssen wir ändern (ggf. Sparce Matrix bauen, d.h. Spalten sind die Autorennamen und Zeilen sind die Filme)
- Country ist eher US-zentriert.
- `date_added` rausschmeißen?
- `release_year` ist interessant (z.B. 80er Filme etc.). Können wir daraus Kategorien bilden? (oder z.B. Filme +/- 5 Jahre)
- `rating` ist relevant
- `duration` müssen wir unterscheiden zwischen Serie und Film.
- `listed_in` müssen wir auch klären (Sparse Matrix wie bei den Schauspielern)
- `Description` - NLP Task?


- Allan Marriott - Film 1
- Allan Marriott - 2013
- Allan Marriott - Comedy
...
- Comedy - Allan Marriott

.drop_duplicates()

# IMDB

In [ ]:
!pip install git+https://github.com/alberanid/imdbpy

  Cloning https://github.com/alberanid/imdbpy to /tmp/pip-req-build-1t7vdwfy
  Running command git clone -q https://github.com/alberanid/imdbpy /tmp/pip-req-build-1t7vdwfy
  Created wheel for IMDbPY: filename=IMDbPY-2020.9.29-cp36-none-any.whl size=302743 sha256=92fc7842da0932bad2403d8ebe2a2245a67250b982181a201004781d69fb809e
  Stored in directory: /tmp/pip-ephem-wheel-cache-cljh3hij/wheels/0f/09/61/190df5e0276765680540f1562f2abca80e725a7e48595e993f
Successfully built IMDbPY


In [ ]:
from imdb import IMDb

# 'Norm of the North: King Sized Adventure'
# create an instance of the IMDb class
ia = IMDb()


In [ ]:
df[df['director'].isna()]

show_id  ...                                        description
2689  80162114  ...  A top detective in the violent crimes unit bec...
4871  80243728  ...  Just months after a tragic loss, detective Sof...
4992  80195357  ...  Immortal renegade Philly the Kid and his trans...
1429  81035120  ...  It's Grabbleapple harvest season in the Rainbo...
5146  80185622  ...  Top medical, criminal and psychological expert...
...        ...  ...                                                ...
2823  80210294  ...  Reggie's wild imagination unlocks a weird and ...
5955  80091742  ...  In a docuseries set at one of NCAA football's ...
3089  81099976  ...  Titipo the train is out to prove that he's got...
4135  81146370  ...  Lucky and friends must figure out how to get h...
405   70303288  ...  Years after a wealthy man is healed by a rural...

[405 rows x 12 columns]

In [ ]:
ia.search_movie('My Little Baby	')

[<Movie id:5546516[http] title:_"My Little Baby" (2016)_>,
 <Movie id:6236592[http] title:_My Little Baby (2019)_>,
 <Movie id:1579638[http] title:_My Little Baby (1916)_>,
 <Movie id:6892950[http] title:_My Little Baby, Jaya (2017)_>,
 <Movie id:6677776[http] title:_"My Little Baby" Episode #1.5 (2016)_>,
 <Movie id:6677778[http] title:_"My Little Baby" Episode #1.6 (2016)_>,
 <Movie id:6677792[http] title:_"My Little Baby" Episode #1.13 (2016)_>,
 <Movie id:6677774[http] title:_"My Little Baby" Episode #1.4 (2016)_>,
 <Movie id:8569206[http] title:_Little Baby (2019)_>,
 <Movie id:7001042[http] title:_Hush Little Baby (2017) (TV)_>,
 <Movie id:10051854[http] title:_"Little Baby Bum" (2011)_>,
 <Movie id:0395140[http] title:_My Little Bride (2004)_>,
 <Movie id:1031294[http] title:_Hush Little Baby (2007) (TV)_>,
 <Movie id:2559620[http] title:_Little Baby Jesus (2013)_>,
 <Movie id:0868745[http] title:_"Teletubbies" Little Baby (1997)_>,
 <Movie id:10741912[http] title:_"Two Dimes" L

In [ ]:
# get a movie
movie = ia.get_movie(5546516)

# print the names of the directors of the movie
movie

<Movie id:5546516[http] title:_"My Little Baby" (2016)_>

In [ ]:
movie.keys()

['original title',
 'cast',
 'genres',
 'countries',
 'country codes',
 'language codes',
 'color info',
 'certificates',
 'number of seasons',
 'rating',
 'votes',
 'cover url',
 'imdbID',
 'languages',
 'title',
 'year',
 'kind',
 'series years',
 'akas',
 'seasons',
 'writer',
 'distributors',
 'plot',
 'canonical title',
 'long imdb title',
 'long imdb canonical title',
 'smart canonical title',
 'smart long imdb canonical title',
 'full-size cover url']

In [ ]:
movie['countries'], movie['cast']

(['Japan'],
 [<Person id:5137121[http] name:_Natsuki Hanae_>,
  <Person id:5190250[http] name:_Risa Taneda_>,
  <Person id:4785118[http] name:_Ayane Sakura_>,
  <Person id:4986657[http] name:_Ryôta Ôsaka_>,
  <Person id:2971661[http] name:_Jennifer Alyx_>,
  <Person id:4285676[http] name:_Megan Crockett_>,
  <Person id:4808080[http] name:_Erika Harlacher_>,
  <Person id:4582537[http] name:_Erik Scott Kimerer_>,
  <Person id:2186567[http] name:_Christina N. McBride_>,
  <Person id:4481399[http] name:_Amanda Céline Miller_>,
  <Person id:7763801[http] name:_Ryan Raydarke_>,
  <Person id:8897900[http] name:_Abelina Sabrina Rios_>,
  <Person id:5982390[http] name:_Adin Rudd_>,
  <Person id:0397364[http] name:_Julie Ann Taylor_>,
  <Person id:6949652[http] name:_Erica Lindbeck_>,
  <Person id:2550198[http] name:_Kyle McCarley_>,
  <Person id:6745380[http] name:_Landon McDonald_>,
  <Person id:5213224[http] name:_Erica Mendez_>,
  <Person id:5489754[http] name:_Max Mittelman_>,
  <Person id:

In [ ]:
movie['director']

KeyError: ignored

In [ ]:
df.loc[4613, :]

IndexError: ignored

In [ ]:
# create an instance of the IMDb class
ia = IMDb()
movies = ia.search_movie('Jandino: Whatever it Takes')
movies

In [ ]:
# get a movie
movie = ia.get_movie(6999080)

# print the names of the directors of the movie
movie.values()

# Graphs

## Networkx

In [ ]:
import networkx as nx

In [ ]:
df['temp_genre']

In [ ]:
df.columns

In [ ]:
df[['temp_genre', 'title']]

In [ ]:
G = nx.from_pandas_edgelist(df, source='temp_genre', target='title')

In [ ]:
import matplotlib.pyplot as plt

plt.figure(figsize=(25, 25))
nx.draw_networkx(G, nodelist=sample_nodes, with_labels=False)

In [ ]:
print(nx.info(G))

In [ ]:
G.edges()

Graphen (Basic):
- Nodes: Filmtitel + Genres

In [ ]:
sorted(nx.connected_components(G))[6]

In [ ]:
titles = df.loc[df[['temp_genre', 'title']].duplicated(), 'title'].values

In [ ]:
df.loc[df['title'].apply(lambda title: title in titles)]

## Graph tool

In [ ]:
!echo "deb http://downloads.skewed.de/apt bionic main" >> /etc/apt/sources.list
!apt-key adv --keyserver keys.openpgp.org --recv-key 612DEFB798507F25
!apt-get update
!apt-get install python3-graph-tool

In [ ]:
import graph_tool.all as gt

# Load table
g = gt.Graph()
# Set property maps for edge attributes
# weight = g.new_edge_property('int')
# color = g.new_edge_property('string')

# Create numpy array of edgelist
edgelist = df[['temp_genre', 'title']].values

# Add edges
node_id = g.add_edge_list(edgelist)

# Access node id of each vertex
for node in range(g.num_vertices()):
    print("Node {} has id: {}".format(node, node_id[node]))

In [ ]:
df[['temp_genre', 'title']].values

# Nächste Schritte

In [ ]:
### BONUS

# GIT
# fastapi
# mlflow
# netwulf
# nlp
# imdb

In [ ]:
df.head(1)

Norm of the King - Richard Finn

Norm of the King - Tim Maltby

...



In [ ]:
def data_clean(features_list[features]):
  ...

  return df # alle Spaltenkombinationen in 'source'/'target', und alle anderen Spalten sind gelöscht

0. Wie komme ich von der aktuellen Datenstruktur zu einem Graphen
1. Funktion bauen, welche einen Titel akzeptiert und ähnliche Titel zurückgibt.










1. Data Cleaning
- Fehlende Werte von `director` entfernen
- `director`, `cast` & `listed_in` liegen aktuell als Listen vor. Endziel wäre pro Zeile eine Filmtitel + Spaltenwert kombination.
- 
2.
3.
4.
5.
6.
7.
